pkl-core/src/main/java/org/pkl/core/service/ExecutorSpiImpl.java

```diff
@@ -98,4 +126,53 @@ public String evaluatePath(Path modulePath, ExecutorSpiOptions options) {
       ModuleKeyFactories.closeQuietly(builder.getModuleKeyFactories());
     }
   }
+
+  private HttpClient getOrCreateHttpClient(ExecutorSpiOptions options) {
+    var clientKey = new HttpClientKey(options);
+    return httpClients.computeIfAbsent(
+        clientKey,
+        (key) -> {
+          var builder = HttpClient.builder();
+          for (var file : key.certificateFiles) {
+            builder.addCertificates(file);
+          }
+          for (var url : key.certificateUrls) {
+            builder.addCertificates(url);
+          }
+          // If the above didn't add any certificates,
+          // builder will fall back to Pkl's built-in certificates.
+          return builder.build();
+        });
+  }
+
+  private static final class HttpClientKey {
+    final Set<Path> certificateFiles;
+
+    @SuppressWarnings("UrlHashCode") // only intended to be used for class path URLs
+    final Set<URL> certificateUrls;
```

Comment:
Switch to `List<URL>`

Reply:
I decided to accept `URL` certificate files in APIs (only) to support loading certificate files from the class path. If we're genuinely concerned about potential side effects of `URL.equals`, I'd suggest to immediately reject URLs that aren't `jar:` or `file:` at API boundaries. Additionally we could use `URI` in APIs and/or internally and only convert to `URL` when accessing the (`jar:` or `file:`) resource. This is probably a worthwhile change; I just wasn't sure if it's necessary.

Reply:
> There's plenty of module keys that have the same entries in both modulesByResolvedUri and modulesByOriginalUri (e.g. file: modules). It's expected for things like this.

OK, I'll revert to the current behavior. That said, conceptually it feels wrong that Pkl treats A and B where A redirects to B as different modules. And if it does, it would seem more consistent to forego a security check of B when accessing A.

Reply:
Not totally sure what you mean, but, it's important that these two are different modules, even if the first URI redirects to the second URI. It would be hard for users to predict the language's behavior otherwise.

```
import "https://example.com/foo.pkl"
import "https://example.com/bar.pkl"
```

Reply:
Moved discussion to #254